packages/remix-server-runtime/single-fetch.ts

```diff
@@ -504,3 +508,76 @@ export function encodeViaTurboStream(
     ],
   });
 }
+
+type MaybePromise<T> = T | Promise<T>;
+
+type Serializable =
+  | undefined
+  | null
+  | boolean
+  | string
+  | symbol
+  | number
+  | Array<Serializable>
+  | { [key: PropertyKey]: Serializable }
+  | bigint
+  | Date
+  | URL
+  | RegExp
+  | Error
+  | Map<Serializable, Serializable>
+  | Set<Serializable>
+  | Promise<Serializable>;
+
+type DataFunctionReturnValue =
+  | Serializable
+  | TypedDeferredData<Record<string, unknown>>
+  | TypedResponse<Record<string, unknown>>;
+
+// Backwards-compatible type for Remix v2 where json/defer still use the old types,
+// and only non-json/defer returns use the new types.  This allows for incremental
+// migration of loaders to return naked objects.  In the next major version,
+// json/defer will be removed so everything will use the new simplified typings.
+// prettier-ignore
+export type Serialize<T extends Loader | ClientLoader | Action | ClientAction> =
+  Awaited<ReturnType<T>> extends TypedDeferredData<infer D> ? D :
+  Awaited<ReturnType<T>> extends TypedResponse<Record<string, unknown>> ? SerializeFrom<T> :
+  Awaited<ReturnType<T>>;
+
+// loader
+type LoaderArgs = RRLoaderArgs<AppLoadContext> & {
+  // Context is always provided in Remix, and typed for module augmentation support.
+  context: AppLoadContext;
+  response: ResponseStub;
+};
+export type Loader = (
+  args: LoaderArgs
+) => MaybePromise<DataFunctionReturnValue>;
+export let defineLoader = <T extends Loader>(loader: T): T => loader;
+
+// clientLoader
+type ClientLoaderArgs = RRLoaderArgs<undefined> & {
+  serverLoader: <T extends Loader>() => Promise<Serialize<T>>;
+};
+type ClientLoader = (args: ClientLoaderArgs) => MaybePromise<Serializable>;
```

Comment:
We also need to support the `clientLoader.hydrate = true` property - here's what we have in the base types:

```ts
type ClientLoaderFunction = ((
  args: ClientLoaderFunctionArgs
) => ReturnType<RRLoaderFunction>) & {
  hydrate?: boolean;
};
```

Reply:
This seemed to be what got it working for me - I needed it in both of those places it seems

```ts
type ClientLoader = ((args: ClientLoaderArgs) => MaybePromise<Serializable>) & {
  hydrate?: boolean;
};
export let defineClientLoader = <T extends ClientLoader>(
  clientLoader: T
): T & { hydrate?: boolean } => clientLoader;
```

Reply:
yeah that crossed my mind too - let's discuss with Ryan if he wants to add that in if they're already making this change now.  We should land on our final preferred API in v2 so that when we stabilize the `v3_singleFetch` flag it's no code changes to v7

Reply:
I like @pcattori idea to pass the hydrate option directly to defineClientLoader, since it's already calling a function it makes sense to just pass the option.